File: Core/Components/MultipleSearchEntry.cs
using Bridge.Html5;
using Core.Clients;
using Core.Components.Extensions;
using Core.Extensions;
using Core.Models;
using Core.MVVM;
using Core.ViewModels;
using System;
using System.Collections.Generic;
using System.Linq;

namespace Core.Components
{
    public class MultipleSearchEntry : SearchEntry
    {
        private const string MultipleClass = "multiple";
        private bool _isStringSource;
        private HTMLButtonElement _toggleButton;

        public MultipleSearchEntry(Component ui) : base(ui)
        {
        }

        public override void Render()
        {
            _isStringSource = Entity != null && Entity.GetType().GetComplexPropType(FieldName, Entity).Equals(typeof(string));
            base.Render();
            Element.ParentElement.AddClass(MultipleClass);
            TryParseData();
            FindMatchText();
        }

        private void TryParseData()
        {
            if (Entity is null)
            {
                return;
            }
            var source = Utils.GetPropValue(Entity, FieldName);
            if (source == null)
            {
                return;
            }
            ICollection<string> list = null;
            if (_isStringSource)
            {
                list = (source as string).Split(",").Where(x => x.HasAnyChar()).ToList();
            }
            else
            {
                list = source.As<ICollection<string>>();
            }
            list.Except(_listValues).ToArray().ForEach(_listValues.Add);
        }

        private readonly List<string> _listValues = new List<string>();

        public List<string> ListValues
        {
            get => _listValues;
            set
            {
                if (value.Nothing())
                {
                    _listValues?.Clear();
                }
                else
                {
                    if (_listValues != value)
                    {
                        _listValues.Clear();
                        value.Distinct().ForEach(_listValues.Add);
                    }
                }

                SetEntityValue(value);
                CascadeField();
                PopulateFields();
            }
        }

        internal override void GridResultDomLoaded()
        {
            base.GridResultDomLoaded();
            if (_toggleButton != null)
            {
                return;
            }
            Html.Take(_gv.Element).Div.ClassName("dropdown-toolbar")
            .Button.Event(EventType.Click, ToggleAllRecord).ClassName("fa fa-check").Attr("title", "Chọn tất cả").End
            .Button.Event(EventType.Click, Dispose).ClassName("fa fa-times").Attr("title", "Đóng gợi ý");
            _toggleButton = Html.Context.PreviousElementSibling as HTMLButtonElement;
        }

        public void ToggleAllRecord()
        {
            var value = ListValues;
            if (value is null)
            {
                ListValues = new List<string>();
                _toggleButton.SetAttribute("title", "Chọn tất cả");
            }
            else
            {
                ListValues = null;
                _toggleButton.SetAttribute("title", "Hủy chọn");
            }
            FindMatchText();
        }

        private void SetEntityValue(ICollection<string> value)
        {
            if (_isStringSource)
            {
                Entity.SetComplexPropValue(FieldName, value.Combine());
            }
            else
            {
                Entity.SetComplexPropValue(FieldName, value);
            }
        }

        public List<object> MatchedItems { get; set; } = new List<object>();

        public override void FindMatchText()
        {
            if (EmptyRow || ProcessLocalMatch())
            {
                return;
            }
            var values = ListValues;
            ClearTagIfNotExists();
            if (MatchedItems.HasElement() && values.Except(MatchedItems.Select(x => x[IdField].As<string>())).Nothing())
            {
                SetMatchedValue();
                return;
            }
            Client.Instance.ComQuery(new SqlViewModel
            {
                ComId = GuiInfo.Id,

            }).Done(ds => {
                MatchedItems = ds.Length > 0 ? ds[0].ToList() : MatchedItems;
                SetMatchedValue();
            });
        }

        protected override bool ProcessLocalMatch()
        {
            var isLocalMatched = _gv != null && RowData.Data.HasElement() || GuiInfo.LocalData != null;
            if (isLocalMatched)
            {
                var rows = GuiInfo.LocalData.Nothing() ? RowData.Data : GuiInfo.LocalData;
                MatchedItems = rows.Where(x => _listValues.Contains(x[IdField]?.ToString())).ToList();
            }
            if (MatchedItems.HasElement() && MatchedItems.Count == _listValues.Count)
            {
                SetMatchedValue();
                return true;
            }
            return false;
        }

        public override void SetMatchedValue()
        {
            _input.Value = string.Empty;
            ClearTagIfNotExists();
            for (var i = 0; i < ListValues.Count; i++)
            {
                var item = MatchedItems.FirstOrDefault(x => x[IdField]?.ToString() == ListValues[i]);
                RenderTag(item);
            }
        }

        private void ClearTagIfNotExists()
        {
            var tags =
                from HTMLElement tag in ParentElement.QuerySelectorAll("div > span")
                let id = tag.Dataset["id"]
                where id != null && !ListValues.Contains(id)
                select tag;
            foreach (var tag in tags)
            {
                tag.Remove();
            }
        }

        private void RenderTag(object item)
        {
            if (item is null)
            {
                return;
            }
            var idAttr = item[IdField].ToString();
            var exist = Element.ParentElement.QuerySelector($"span[data-id='{idAttr}']");
            if (exist != null)
            {
                return;
            }
            Html.Take(Element.ParentElement).Span.Attr("data-id", idAttr).InnerHTML(GetMatchedText(item));
            var tag = Html.Context;
            Element.ParentElement.InsertBefore(Html.Context, _input);
            Html.Instance.Button.ClassName("fa fa-times").End.Event(EventType.Click, () =>
            {
                var oldList = ListValues.ToList();
                MatchedItems.Remove(item);
                var id = item[IdField]?.ToString();
                while (ListValues.Contains(id))
                {
                    ListValues.Remove(id);
                }
                ListValues = ListValues;
                Dirty = true;
                FindMatchText();
                UserInput?.Invoke(new ObservableArgs { NewData = ListValues, OldData = oldList, EvType = EventType.Change });
                tag.Remove();
                this.DispatchEvent(GuiInfo.Events, EventType.Change, Entity, ListValues, oldList).Done();
            }).End.Render();
        }

        protected override void EntrySelected(object rowData)
        {
            Window.ClearTimeout(_waitForDispose);
            if (rowData is null)
            {
                return;
            }

            var id = rowData[IdField]?.ToString();
            if (ListValues is null)
            {
                ListValues = new List<string>();
            }
            if (!ListValues.Contains(id))
            {
                ListValues.Add(id);
                MatchedItems.Add(rowData);
            }
            else
            {
                ListValues.Remove(id);
                var exist = MatchedItems.FirstOrDefault(x => x[IdField]?.ToString() == id);
                MatchedItems.Remove(exist);
            }
            ListValues = ListValues;
            Dirty = true;
            FindMatchText();
            _input.Focus();
            UserInput?.Invoke(new ObservableArgs { NewData = ListValues, OldData = ListValues, NewMatch = rowData, EvType = EventType.Change });
            this.DispatchEvent(GuiInfo.Events, EventType.Change, Entity).Done();
        }

        public override void UpdateView(bool force = false, bool? dirty = null, params string[] componentNames)
        {
            TryParseData();
            FindMatchText();
        }
    }
}

File: frontend/lib/multipleSearchEntry.js
import { SearchEntry } from "./searchEntry.js";
import { Client } from "./clients/client.js";
import { Html } from "./utils/html.js";
import { Utils } from "./utils/utils.js";
import EventType from "./models/eventType.js";

export class MultipleSearchEntry extends SearchEntry {
    static multipleClass = "multiple";
    _toggleButton = null;
    isMultiple = true;

    constructor(ui) {
        super(ui);
    }

    render() {
        this._listValues = [];
        this.setDefaultVal();
        this.tryParseData();
        this.renderInputAndEvents();
        this.findMatchText();
        this.searchResultEle = document.body;
        this.element.parentElement.classList.add(MultipleSearchEntry.multipleClass);
        this.element.parentElement.addEventListener("click", () => {
            this._input.focus();
        });
    }

    tryParseData() {
        if (!this.entity) {
            return;
        }
        let source = this.entity[this.Name];
        if (!source) {
            this.Matched = null;
            this._listValues = [];
            return;
        }
        this._listValues = source.toString().split(this.meta.groupFormat || ',').filter(x => x.trim().length > 0);
    }

    _listValues = [];

    get listValues() {
        return this._listValues;
    }

    set listValues(value) {
        if (!value) {
            this._listValues = [];
        } else {
            this._listValues = Array.from(new Set(value));
        }
        this.setEntityValue();
    }

    setEntityValue() {
        this.entity[this.Name] = this.listValues.length > 0 ? this.listValues.join(this.meta.groupFormat || ',') : null;
        this.entity[this.Name + "Text"] = this.matchedItems.length > 0 ? this.matchedItems.map(item => this.getMatchedText(item)).join(this.meta.groupFormat || ',') : this.entity[this.Name + "Text"];
    }

    matchedItems = [];

    findMatchText() {
        if (this.emptyRow) {
            return;
        }
        if (!this.processLocalMatch()) {
            this.setMatchedValue();
        }
    }

    processLocalMatch() {
        if (Utils.isNullOrWhiteSpace(this.meta.refName)) {
            var data = Utils.isFunction(this.meta.Query, false, this);
            this.matchedItems = data.filter(x => this.listValues.includes(x.Id.toString()));
            this.setMatchedValue();
            this.entity[this.Name + "Text"] = this.matchedItems.length > 0 ? this.matchedItems.map(item => this.getMatchedText(item)).join(this.meta.groupFormat || ',') : this.entity[this.Name + "Text"];
            return true;
        }
        else {
            this.Matched = this.entity[this.displayField] || null;
            if (this._listValues.length > 0 && this.matchedItems.filter(x => this._listValues.includes(x.Id)).length < this._listValues.length && (!this.parent.isListViewItem || this.meta.isMultiple)) {
                Client.instance.getByIdAsync(this.meta.refName, this._listValues).then(data => {
                    this.matchedItems = data.data ? data.data : [];
                    if (this.matchedItems.length != this._listValues.length) {
                        this.listValues = this.matchedItems.map(x => x[this.idField].toString());
                    }
                    this.setMatchedValue();
                    this.entity[this.Name + "Text"] = this.matchedItems.length > 0 ? this.matchedItems.map(item => this.getMatchedText(item)).join(this.meta.groupFormat || ',') : this.entity[this.Name + "Text"];
                })
                return true;
            }
        }
        return false;
    }

    setMatchedValue() {
        this._input.value = '';
        this.listValues.forEach(value => {
            let item = this.matchedItems.find(x => x[this.idField].toString() === value.toString());
            this.renderTag(item);
        });
    }

    clearTagIfNotExists() {
        Array.from(this.element.parentElement.querySelectorAll("span")).forEach(ta => {
            ta.remove();
        });
    }

    renderTag(item) {
        if (!item) {
            return;
        }
        let idAttr = item[this.idField];
        let exist = this.element.parentElement.querySelector(`span[data-id='${idAttr}']`);
        if (exist) {
            return;
        }
        Html.take(this.element.parentElement).span.attr("data-id", idAttr).i.className("fal fa-tag mr-1").end.text(this.getMatchedText(item));
        var tag = Html.context;
        this.element.parentElement.insertBefore(Html.context, this._input);
        if (this.disabled) {
            this._input.readOnly = true;
        }
        Html.instance.button.className("fa fa-times").event(EventType.Click, async () => {
            if (this.disabled) {
                return;
            }
            let oldMatch = this.matchedItems;
            this.matchedItems.splice(this.matchedItems.indexOf(item), 1);
            var id = item[this.idField];
            this.listValues = this.listValues.filter(x => x !== id.toString());
            this.setEntityValue();
            this.Dirty = true;
            if (this.userInput != null) {
                this.userInput?.invoke({ newData: this._value, oldData: oldMatch, evType: EventType.Change });
            }
            await this.dispatchEvent(this.meta.events, EventType.Change, this);
            tag.remove();
        }).end.render();
    }

    entrySelected(rowData) {
        window.clearTimeout(this._waitForDispose);
        this.emptyRow = false;
        if (rowData === null || this.disabled) {
            return;
        }

        let oldMatch = this.matchedItems;
        var id = rowData[this.idField];
        if (this.listValues.length == 0 || !this.listValues.includes(id)) {
            this.listValues.push(id.toString());
            this.matchedItems.push(rowData);
        }
        else {
            return;
        }
        this.setEntityValue();
        this.Dirty = true;
        this.findMatchText();
        this._gv.allListViewItem.forEach(item => {
            item.setChooseCell();
        });
        if (this.userInput != null) {
            this.userInput?.invoke({ newData: this._value, oldData: oldMatch, evType: EventType.Change });
        }
        this.dispatchEvent(this.meta.events, EventType.Change, this).then();
    }

    updateView(force = false, dirty = null, ...componentNames) {
        this.tryParseData();
        this.setEntityValue();
        this.clearTagIfNotExists();
        this.findMatchText();
    }
}
